warp/Network/Wai/Handler/Warp/HTTP2/Receiver.hs

```diff
@@ -0,0 +1,307 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+
+module Network.Wai.Handler.Warp.HTTP2.Receiver (frameReceiver) where
+
+#if __GLASGOW_HASKELL__ < 709
+import Control.Applicative
+#endif
+import Control.Concurrent (takeMVar)
+import Control.Concurrent.STM
+import qualified Control.Exception as E
+import Control.Monad (when, unless, void)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import Data.Maybe (isJust)
+import Network.HTTP2
+import Network.HTTP2.Priority
+import Network.Wai.Handler.Warp.HTTP2.EncodeFrame
+import Network.Wai.Handler.Warp.HTTP2.HPACK
+import Network.Wai.Handler.Warp.HTTP2.Request
+import Network.Wai.Handler.Warp.HTTP2.Types
+import Network.Wai.Handler.Warp.IORef
+import Network.Wai.Handler.Warp.Types
+
+----------------------------------------------------------------
+
+frameReceiver :: Context -> MkReq -> (BufSize -> IO ByteString) -> IO ()
+frameReceiver ctx@Context{..} mkreq recvN =
+    E.handle sendGoaway loop `E.finally` takeMVar wait
+  where
+    sendGoaway (ConnectionError err msg) = do
+        csid <- readIORef currentStreamId
+        let frame = goawayFrame csid err msg
+        enqueue outputQ (OGoaway frame) highestPriority
+    sendGoaway _                         = return ()
+
+    sendReset err sid = do
+        let frame = resetFrame err sid
+        enqueue outputQ (OFrame frame) highestPriority
+
+    loop = do
+        hd <- recvN frameHeaderLength
+        if BS.null hd then
+            enqueue outputQ OFinish highestPriority
+          else do
+            cont <- processStreamGuardingError $ decodeFrameHeader hd
+            when cont loop
+
+    processStreamGuardingError (_, FrameHeader{..})
```

Comment:
when there are multiple `Context{...}` and `FrameHeader{..}` it is a bit hard to know where `streamId`, `continued` etc are coming from.   Suggest `FrameHeader{streamId}`.